### ISSUE TITLE
Error to access notifications and hide comments

### ISSUE THREAD

**Issue by sergiochtf** [consul/consul#1581]
I have problems for to view notifications of my account and hide comments in admin panel.  I attach 2 screens.
![p1](https://cloud.githubusercontent.com/assets/9718722/26619881/5de84efa-45d8-11e7-8f3c-b6fd0c42598f.png)
![p2](https://cloud.githubusercontent.com/assets/9718722/26619880/5de205c2-45d8-11e7-9bfb-bc3519c605a4.png)

**Comment by bertocq** [comment]
Hello @sergiochtf 

For the first one, seems like you've a notification without a notifiable object. That's where the `undefined method for nil:NilClass` comes from. Can you check it to confirm?

For the second, I get a similar error but for a comment that has a Poll Question as commentable, because there is no path/view to link to. In your case seems like you've a comment for a Budget and there is no GET /management/budgets/:id route. This one needs a bit more of research, I'll try first to figure out the error I'm seen and then I'll try to reproduce yours.

**Comment by sergiochtf** [comment]
Hi @bertocq, for the first one, I check the object and this exist en DDBBB. For the second, I wait your comments.

**Comment by bertocq** [comment]
@sergiochtf debugging in remote is hard, could you post some more details about the first one?
I see complete code around the error is:
```
  def notifiable_title
    case notifiable.class.name
    when "ProposalNotification"
      notifiable.proposal.title
    when "Comment"
      notifiable.commentable.title
    else
      notifiable.title
    end
  end
```
So getting the class of the notification that's giving the error may be the clue to solve it.

**Comment by sergiochtf** [comment]
@bertocq i put the data of the error:
`{:action=>"show"
:budget_id=>#<Budget::Investment id: 117
author_id: 62
administrator_id: nil
title: "prueba"
description: "<p>a</p>\r\n"
external_url: ""
price: nil
feasibility: "feasible"
price_explanation: ""
unfeasibility_explanation: ""
internal_comments: ""
valuation_finished: true
valuator_assignments_count: 2
price_first_year: nil
duration: ""
hidden_at: nil
cached_votes_up: 0
comments_count: 2
confidence_score: 1
physical_votes: 0
tsv: "'centr':6B 'centro-if':5B 'distrit':4B 'ifar':7B '..."
created_at: "2017-04-27 15:02:04"
updated_at: "2017-04-27 16:08:51"
heading_id: 62
responsible_name: "00000000"
budget_id: 11
group_id: 22
selected: true
location: nil
organization_name: nil
unfeasible_email_sent_at: nil
ballot_lines_count: 1
previous_heading_id: nil
winner: false>
:controller=>"budgets/investments"
:id=>nil}`

**Comment by bertocq** [comment]
@sergiochtf I was asking about debugging the value when the first error happens.

Like, when `notifiable.title` is executed `notifiable` is nil as the first screenshot tells. Thus I wonder what the Notification object looks like. Can you add a `byebug` line at the top of the method, refresh the browser and see on the server console what it prints when you type `self` ?

**Comment by sergiochtf** [comment]
@bertocq what file you need that I put the line byebug?

**Comment by sergiochtf** [comment]
@bertocq the file is app/models/notification.rb ??

**Comment by sergiochtf** [comment]
![3](https://cloud.githubusercontent.com/assets/9718722/26633037/0181518e-460a-11e7-9586-fc20e031247e.png)

**Comment by bertocq** [comment]
@sergiochtf I meant that `byebug` to be at the top of the method, not the class, like:
```
 def notifiable_title
   byebug if notifiable.nil?

    case notifiable.class.name
    when "ProposalNotification"
      notifiable.proposal.title
    when "Comment"
      notifiable.commentable.title
    else
      notifiable.title
    end
  end
```

That `   byebug if notifiable.nil?` will open a debug console on the terminal where you're executing the server, only for the scenario I suspect is giving you the error: " A notification without a notifiable element associated". Its just so you can confirm my theory, since looking at the database seemed not to be conclusive.

Assuming I'm right and the `notifiable` is `nil` (not present) for one notification, I wonder how your database got to that state:

A- You somehow where touching records throught the rails console (or the database) and managed to create a Notification without a notifiable, or removed the relationship between the two. That would be the easiest explanation and no fix could be done further than totally removing the Notification object or creating again the relationship (if the notifiable still exists)

B- You somehow managed to create a Notification without a notifiable through the interface. That would be a bug and we should find out how you did it, to fix that bug in the UI.

Anyway I would assume that all Notifications should always have a notifiable associated (is that assumption right @xuanxu @voodoorai2000 ??) and create a PR to avoid any record in that state sleeping up to the interface to avoid the UI from breaking.

**Comment by bertocq** [comment]
@sergiochtf I understand what you're saying is that reason behind error 1 was a manual change in the database? Okey then :).

About error 2 (that I assume its https://cloud.githubusercontent.com/assets/9718722/26619881/5de84efa-45d8-11e7-8f3c-b6fd0c42598f.png) I get from your last response that you deleted the debate where the comment was made, that could be the reason behind. Any manual change on the database, or through the rails console, always bypass the rails models restrictions/verifications and for sure may leave the database in an invalid state that may throw errors. Its always nice to make the code expect those invalid scenarios anyway, assuming the worst, but that can be seen as too much defensive programming.

**Comment by sergiochtf** [comment]
Ok, I understand you, but then, What remove comments, debates, etc from the application? I don't view any option for this.

**Comment by bertocq** [comment]
I'm not sure why comments, debates and other entities don't have a "delete" action on the admin panel. I guess its a product requirement spec but its just a guess, I've very few knowledge about consul (I've been on this codebase/product for a week only). Maybe @voodoorai2000 or @xuanxu can give info about it (if it's a feature to come, or there is a reason behind it)

On your particular error, yes seems like given the text from rails, there is a Budget with an invalid state (or relationships) that is the reason behind the error. If you want to get pass it, just removing it totally from database and any related object, and creating it again in case you need it, its going to be your way out.

I would be worried about this errors appearing because of user actions (because of actions that could be performed through the UI) but touching the database manually its for sure not supported :), maybe it's time to create a Pull Request to add those "removal" actions to the entities you want to cleanup and restrict that action just to Admins (or even allow to enable/disable it totally through Setting variable) What do you think?

**Comment by bertocq** [comment]
@sergiochtf instead of "deleting".. have you tried the "hidding" feature for proposals/comments/debates/users ? It acts as if the content was deleted, but it can be restored just in case...

![screen shot 2017-05-31 at 19 13 11](https://cloud.githubusercontent.com/assets/983242/26644516/451ffd48-4635-11e7-9100-1bc37f03433b.jpg)

**Comment by xuanxu** [comment]
There is no option to delete content. The idea is that the software is going to be used by governments and institutions and removed content should be auditable. So the option that is available is to hide content. Moderators and admins have extra contextual menus to hide content and block users:

![ocultar](https://cloud.githubusercontent.com/assets/6528/26676413/fe7309a6-46c7-11e7-8cad-ad6f4930b999.png)

**Comment by sergiochtf** [comment]
For the error of the comment, clean the database is not possible, too much data loaded. ¿Any solution? I tolk to the administrator for not to delete data from database. The comment, notification and budget_investment exists in your tables.

**Comment by voodoorai2000** [comment]
@sergiochtf Indeed you have found an exception that we've been meaning to solve for a while.
Notifications for comments on investments do not generate the route correctly and throw the error in your first screenshot. Maybe using this [helper](https://github.com/consul/consul/blob/master/app/helpers/comments_helper.rb#L23) will fix it.

Regarding the second situation, it looks like you have hidden a notifiable and due to the paranoia gem, the notifiable is not found. We should add a spec and make the code aware of this possible situation too.

Regarding your situation with the existing DB records, it should be solved by updating the code appropriately with these 2 PRs needed. Temporarily maybe finding the appropriate hidden record and setting the hidden_at attribute to nil would solve it. [Here](https://github.com/rubysherpas/paranoia#usage) is more info on paranoia and how to find hidden records.

**Comment by sergiochtf** [comment]
Ok @voodoorai2000 , but how I use the [helper ](https://github.com/consul/consul/blob/master/app/helpers/comments_helper.rb#L23)to fix it? And the second error, access to hidden comments, when a I change the loggin user, in panel moderation, not show error. With my account of admin show the error, but when I use the account admin by default, it is not show. The error show when I click to link "All". What could be happening?

![p1](https://cloud.githubusercontent.com/assets/9718722/26778713/57fa35e8-49da-11e7-9a49-5b07e1c7b459.png)

**Comment by voodoorai2000** [comment]
@sergiochtf to use the helper it would be something like this:
`<%= link_to comment.commentable.title, commentable_path(comment) %>`

Regarding the other issue, I suggest we create a new issue, just to deal with that.
If you could provide reproducible instructions we'll try it out locally and figure it out.
A good starting point could be the dev_seeds, from there we can create a common data set to reproduce the error locally.

**Comment by sergiochtf** [comment]
Hi @voodoorai2000: With this change <%= link_to comment.commentable.title, commentable_path(comment) %>, I resolved my error. If you like, open a new issue for talk about the notification error link, if  just you don't have any for this error.

**Comment by voodoorai2000** [comment]
Awesome!

We should solve this in Consul so that other people don't run into this problem.
Let's leave this issue open until we have a PR to fix it.

@sergiochtf if you would like to give it a shot, the spec would be similar to [this one](https://github.com/consul/consul/blob/master/spec/features/moderation/comments_spec.rb#L48)

And [here](https://github.com/consul/consul/blob/master/CONTRIBUTING_EN.md) you can find the contributing guidelines

Cheers!

**Comment by sergiochtf** [comment]
I don't know if it is correct, but here is the link:

https://github.com/consul/consul/pull/1611
https://github.com/consul/consul/pull/1610

**Comment by javierm** [comment]
Closed via #2395.
